Use ajax power on module order field

dgrammatiko opened the pull request:
#### Enhance the current status quo

Try to create a new module in module manager. Select a position. Select another position
The dropdown ordering doesn't really represent the true order of the modules in the new position.
So my idea was to use ajax power to be always accurate on that dropdown.
#### Summary of Changes

Use ajax power, tight connection with the controller module.php
#### Preview

Mind the untranslated string!
No other module in that position:
![screen shot 2016-07-06 at 23 23 04](https://cloud.githubusercontent.com/assets/3889375/16633245/05758102-43d1-11e6-80c0-6bcbdd2c1aa6.png)

With other modules in the same position:
![screen shot 2016-07-06 at 23 24 48](https://cloud.githubusercontent.com/assets/3889375/16633259/177b2f5a-43d1-11e6-9c3c-d6c5fa834c79.png)
#### Testing

Please DO NOT TEST this one!
Instead give a vote (up or down thumb) so PLT can make a decision (or PLT decides directly, whatever comes first)

dgrammatiko committed:
rfc

dgrammatiko committed:
CS

Devportobello commented on libraries/cms/form/field/moduleorder.php:
```diff
@@ -44,55 +43,51 @@ protected function getInput()
+							"position": originalPos
+					 },
+					success:function(response) {
+					 	console.log(response.data);
```
debug code ?

dgt41 replied:
This PR is just to showcase some working code. it's not commit ready. 
Eg `$("#jform_position")` the relation with the position field right now is hardcoded, needs an option in the XML for the field moduleorder

Devportobello commented on media/system/js/core.js:
```diff
@@ -1 +1,636 @@
-Joomla=window.Joomla||{},Joomla.editors=Joomla.editors||{},Joomla.editors.instances=Joomla.editors.instances||{},function(e,t){"use strict";e.submitform=function(e,n,o){n||(n=t.getElementById("adminForm")),e&&(n.task.value=e),n.noValidate=!o,n.setAttribute("novalidate",!o);var r=t.createElement("input");r.style.display="none",r.type="submit",n.appendChild(r).click(),n.removeChild(r)},e.submitbutt ...
```
The following code are not minified, so this is for core-uncompressed.js :)

Devportobello commented on administrator/components/com_modules/controllers/module.php:
```diff
@@ -224,4 +224,80 @@ public function save($key = null, $urlVar = null)
+
+				$orders2[$orders[$i]->position]++;
+				$ord = $orders2[$orders[$i]->position];
+				$title = JText::sprintf('COM_MODULES_OPTION_ORDER_POSITION', $ord, addslashes($orders[$i]->title));
```
RFC i know, just to advice: addslashes give me: "11. Who\'s Online"

dgt41 replied:
That was a blind transfer of the old code, nice finding I will fix that!

Devportobello commented:
Dunno if people use the ordering directly in edit mode, but this is an improvment when there is many module in same position

dgt41 commented:
@Devportobello from a UX perspective this is the right approach.
Now is this something that people actually use? I have no clue...

RonakParmar commented:
I have tested this item :red_circle: unsuccessfully on 4885e07250ca96c2dfce70705b733b1119a22ec3<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11040">issues.joomla.org/joomla-cms/11040</a>.</sub>

RonakParmar commented:
I have applied this PR and try to create "custom" module, I have jQuery error in console.
Joomla Version: Joomla! 3.6.0-rc Release Candidate [ Noether ] 28-June-2016 20:34 GMT 
Tested in Ubuntu 12.04, FireFox 47.0.1.![screen shot 2016-07-12 at 07 15 42](https://issues.joomla.org/uploads/1/0d88021f8d6e7562f3ea12e1ec98692a.png)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11040">issues.joomla.org/joomla-cms/11040</a>.</sub>

killoltailored commented:
I have tested this item :red_circle: unsuccessfully on 4885e07250ca96c2dfce70705b733b1119a22ec3<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11040">issues.joomla.org/joomla-cms/11040</a>.</sub>

killoltailored commented:
I have tested after applying patch with Joomla Version Joomla! 3.6.0-rc Release Candidate [ Noether ] 28-June-2016 20:34 GMT in Windows 8.1, Firefox 47.0.1
Create new custom module and found jQuery error. Test with both case select unused position and used position.![screen shot 2016-07-12 at 07 27 03](https://issues.joomla.org/uploads/1/2ffd10d39ed1a5a8088a66737ed02dba.png)![screen shot 2016-07-12 at 07 27 06](https://issues.joomla.org/uploads/1/f4bc16c35c93047b5bea318fb81ccbb9.png)<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11040">issues.joomla.org/joomla-cms/11040</a>.</sub>

dgt41 commented:
@RonakParmar @killoltailored Thanks for testing but this is RFC (Request for comments) which means a decision is needed here before I spent more time on this pr. So an up vote or down vote on the scope is enough at the moment (even if the code is not show time ready). Thanks again

dgt41 commented:
@brianteeman I thought that RFC indicator was enough. Maybe I should have opened this as an issue without code attached so people wouldn't be able to do any testing.

brianteeman commented:
Don't forget that we have a lot of less experienced developers and new
testers.

Never assume anything - when you assume you make a s ASS out of U and ME

gunjanpatel commented:
I am sure they are testing it because they see `pending` status of the task. It is hard to understand what is RFC from the title of the task. :smile:  

It isn't should be in a "Needs Review" status? Or may just "New" status and which waiting to be "Confirmed". 

Thanks.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11040">issues.joomla.org/joomla-cms/11040</a>.</sub>

brianteeman commented:
Gunman. Unfortunately any PR will show up in patch tester. Patchtester only
shows if it is already RTC or not. Needs Review is an internal flag for CMS
maintainers to make a decision on something.

To be honest a comment that it didn't work is a reply to a request for a
comment in my book :)

JoshuaLewis commented:
Great idea. So who are you asking if this is worth it before working on this PR? The folks who manage PR's for new releases of Joomla? Testers?

brianteeman commented:
@dgt41 - seems the consensus is that yes this is a good idea but that the code currently doesnt work yet<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11040">issues.joomla.org/joomla-cms/11040</a>.</sub>

dgrammatiko committed:
Fixes:

- Seperate html-javascript (no inline scripts)
- Fix broken ajax
- Fix wrong escaping
- Introduce option for the linked field
- No new strings

dgt41 commented:
Now it should be testable

dgrammatiko committed:
core.js

gunjanpatel commented:
Can you remove RFC from title then?

brianteeman commented:
Not sure if it is doing what I expect it to do
After applying the patch the ordering in the select is correct but when viewing it on the front end the ordering selected is not being used<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11040">issues.joomla.org/joomla-cms/11040</a>.</sub>

dgt41 commented:
@brianteeman before applying this patch was the behaviour different? All this pr is doing is reloading the ordering when the position is changed, it shouldn't interfere with the actual ordering in the front end...

andrepereiradasilva commented on libraries/cms/form/field/moduleorder.php:
```diff
@@ -19,80 +19,123 @@ class JFormFieldModuleOrder extends JFormField
+	 *
+	 * @note    The field does not include an upload mechanism.
+	 * @see     JFormFieldMedia
+	 * @since   11.1
```
`11.1`? better always to use `__DEPLOY_VERSION__`

gunjanpatel replied:
I think `__DEPLOY_VERSION__` needs to be on other places too where `3.7` is added.

dgt41 replied:
@andrepereiradasilva @gunjanpatel you are both right but I think `__DEPLOY_VERSION__` came after july 6th. Anyways I will fix this..

gunjanpatel replied:
ohh sorry didn't noticed dates. Thanks :+1:

brianteeman commented:
I have tested this item :white_check_mark: successfully on 9ed547eb37f514d12c3c03859a1f4d18d75a6f40<br /><br />No idea what the issue was I saw before as I cant replicate it now

I still dont really see what the difference is now though compared to before this pr<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11040">issues.joomla.org/joomla-cms/11040</a>.</sub>

AnishaTailored commented:
I have tested this item :white_check_mark: successfully on 9ed547eb37f514d12c3c03859a1f4d18d75a6f40<br /><br />I have tested this successfully. If we change the position then the ordering field fetches modules of the newly selected position.
But shouldn't Ordering dropdown be jQuery chosen dropdown as it was before this PR.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11040">issues.joomla.org/joomla-cms/11040</a>.</sub>

zero-24 commented:
@dgt41 can you double check the since tags? If that is done i think we can RTC

dgrammatiko committed:
use Deploy_version

dgt41 commented:
@zero-24 there you go!

zero-24 commented on libraries/cms/form/field/moduleorder.php:
```diff
@@ -19,80 +19,119 @@ class JFormFieldModuleOrder extends JFormField
+	 *
+	 * @return  string  The field input markup.
+	 *
+	 * @since   11.1
```
I think this is missing?

dgt41 replied:
getInput() was there before so the since should not change. (11.1 is wrong should be 1.6)

zero-24 commented on libraries/cms/form/field/moduleorder.php:
```diff
@@ -19,80 +19,119 @@ class JFormFieldModuleOrder extends JFormField
+	 * @return  boolean  True on success.
+	 *
+	 * @see     JFormField::setup()
+	 * @since   3.2
```
And this one?

zero-24 commented on administrator/components/com_modules/controllers/module.php:
```diff
@@ -224,4 +224,80 @@ public function save($key = null, $urlVar = null)
+		{
+			$html[] = $position . ',' . 1 . ',' . JText::_('JNONE');
+		}
+
```
Here is just one space to much ;)

zero-24 commented:
Hmm i have found two more ;)

dgrammatiko committed:
since

dgrammatiko committed:
remove extra line

zero-24 commented:
Go for it @joomla-cms-bot please make us happy :smile:<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/11040">issues.joomla.org/joomla-cms/11040</a>.</sub>
